### DISCUSSION TITLE
check the conservation of angular momentum

### DISCUSSION THREAD

**Discussion by yzd0014** [google-deepmind/mujoco#1451] [General]
I'm not sure if mojoco keeps track the total angular momentum of all bodies with respect to a reference point just like energy. I'm not sure if subtree_angmom will give what I want.

**Comment by yuvaltassa** [comment]
Yes, subtree_angmom is what you want. Note that this computation is optional. You need to add the appropriate sensor to your model if you want this to computed automatically.

**Reply by yzd0014** [reply]
Thank you for your reply! Correct me if I'm wrong, in order to choose a reference point for angular momentum I need to choose a fixed body as the root of subtree. The most reasonable one is the world body, but what I realized is that I can't have a name for world body. Or I need to add a dummy body or something? I'm not sure if my question makes sense.

**Reply by yuvaltassa** [reply]
1. The name of the world body is "world".
2. You cannot choose the reference anyway, it is always the CoM of the subtree of the body. Hence `subtree_angmom` which is a `nv x 3` matrix.

**Reply by yzd0014** [reply]
Thank you for the clarification. So for the 2nd part of your reply, I'm actually really confused. 
1. In documentation, it says subtree_angmom is nbodyx3. My original understanding is that, it stores angular momentum for each rigid body, and the reference point is the COM of the subtree, is this right? 
2. In order to check angular momentum conservation, I need to make sure reference point is not changing over time. I set the sensor so that the root of subtree is "world". I consider world has infinite mass, thus COM of subtree doesn't change. Can I understand it in this way?
3. With double pendulum as an example, I have three bodies, one world, two for double pendulum. I printed the first three values of subtree_angmom, it's changing overtime. This doesn't makes sense to me since I think the world body is not moving at all. I think my understanding for what subree_angmom stores is off.

**Reply by yuvaltassa** [reply]
1. This is correct.
2. "In order to check angular momentum conservation, I need to make sure reference point is not changing over time." This is true for anchored bodies but not for free bodies.
3. "I consider world has infinite mass, thus COM of subtree doesn't change. Can I understand it in this way?" No. The mass of the world is (chosen to be) zero, the com of the world is actually the com of all its children.
4. Okay this is a tricky consequence of Noether's theorem. The conserved angmom for a pendulum is indeed wrt to the anchor, not wrt the com, but MJ computes angmom wrt the com. Basically subtree_angmom is designed for free subtrees, not for anchored subtrees.

I wonder if @v-r-a has anything valuable to add to this, as he's been looking at the angmom code recently.

**Reply by yuvaltassa** [reply]
I think (have not tested) that to get the angmom wrt the root, you can take the angmom wrt the com and add the com linvel x subtree_mass. This should be conserved w/o external forces.

I hope to try this later today, will let you know.

**Reply by yzd0014** [reply]
Thank you very much, I understand it much better now. So correct me if I'm wrong.
1. If the root joint is a free joint, then I can use subtree_angmom to check conservation of angular momentum, right? I originally treat anchored cases the same as free cases. I consider anchor point as a particle that has infinite mass so that it's static and the anchor point is also COM. But currently subtree_angmom doesn't support this logic.
2. Regarding your suggestions for anchored case. I think I can use subtree_angmom to compute angular momentum with COM first. Then I will add position of COM x (mass * linvel of COM). Also I think this will only work if I have anchor at origin so that the position of COM can be directly used as moment arm.
3. For the double pendulum case, if I want the COM of the double pendulum, do I set the root body to be "world" or the first link of the double pendulum for subree_angmom. I'm kind of confused about the definition.

**Reply by yuvaltassa** [reply]
1. Yes.
2. Yes, I meant the vector from the anchor to the com. I haven't tested this, but think it will work.
3. It will be the same if there is only one pendulum in the world. If there are 2 the subtree_com of the world will be between them. Makes sense?

**Reply by yzd0014** [reply]
Thanks for confirming it. For my 3rd question, after I provide a body name in the sensor for subtree_com, I'm not sure if COM is computed for this body and all its children or only all its children. According to your answer, it seems like I can use any body in a subtree for the sensor and it doesn't have to the root of that subtree? With double pendulum, I can even have the body name of the 2nd link for "name" under sensor/⁠subtreeangmom in the xml file, the COM is still computed for the entire double pendulum?

**Reply by v-r-a** [reply]
Hello, 
I'm a student and MuJoCo user. Here are a few details as I understand:

> I'm not sure if mojoco keeps track the total angular momentum of all bodies with respect to a reference point just like energy. I'm not sure if subtree_angmom will give what I want.

You will have to calculate the angular momentum of each body (about its own COM ($h_g$) to start with) yourself. It is not readily available anywhere. But [this](https://github.com/google-deepmind/mujoco/blob/e77c3cb2581dd1f45e30d50b31cb0a88de26a626/src/engine/engine_core_smooth.c#L1693) is the place where it is calculated in the [subtree_angmom code](https://github.com/google-deepmind/mujoco/blob/e77c3cb2581dd1f45e30d50b31cb0a88de26a626/src/engine/engine_core_smooth.c#L1713).

The quantity readily available is the angular momentum of a chain of bodies (aka subtree) about the subtree's centre of mass. If you access `subtree_angmom`, you will get the angular momentum of the whole robot about its COM. 
If you access `subtree_angmom +3*bodyid` you will get the angmom of the rest of the robot start from `bodyid` about the rest of the robot's COM.

(Just reiterating that) there is no difference in the angmom value if your bodyid is 0 or 1 if there is only one tree/robot in the model.

As @yuvaltassa has rightly mentioned, $h_p = h_g + (r_g - r_p) \times (m v_g)$ for a body, where $h_p$ is the angmom about point P fixed w.r.t. the world frame.

**Reply by yuvaltassa** [reply]
> I'm not sure if COM is computed for this body and all its children or only all its children. 

The former: the body and all its children. Because the world has no momentum itself, if there is only one subtree in the model, it doesn't matter if you choose the world body or the root body. If you choose a leaf body you will get something else (basically the angmom of the leaf body about its own com). Makes sense?

**Reply by yzd0014** [reply]
@v-r-a Thank you very much for the info you provided!

**Reply by yzd0014** [reply]
@yuvaltassa Yes, it makes sense. By reading the reply from @v-r-a, I think I have wrong understanding previously. Please let me know if this is correct. Previously I thought every three entries in subtree_angmom stores angular momentum of a rigid body wrt com, but it actually stores angular momentum of a rigid and all its children?

**Reply by yzd0014** [reply]
I was able to verify the conservation of angular momentum by adding the extra term as you suggested. Thank you for your help! @v-r-a @yuvaltassa
